File: apps/backend/scripts/predev_check.py
#!/usr/bin/env python
"""predev guard: refuses to start a second `uvicorn` for THIS PostPiloter
backend if one is already running, mirroring apps/frontend/scripts/predev-check.js.

Detects by process command line (backend directory path + "uvicorn"), not by
port, since a duplicate instance can end up on a different port. Never kills
anything. If detection itself fails, it fails open.
"""

import subprocess
import sys
from pathlib import Path

if sys.stdout.encoding and sys.stdout.encoding.lower() != "utf-8":
    sys.stdout.reconfigure(encoding="utf-8")

BACKEND_DIR = str(Path(__file__).resolve().parent.parent)


def find_running_instances():
    escaped_dir = BACKEND_DIR.replace("'", "''")
    ps_script = f"""
    $ErrorActionPreference = 'Stop'
    $dir = '{escaped_dir}'
    $procs = Get-CimInstance Win32_Process -Filter "Name='python.exe'" |
      Where-Object {{
        $_.CommandLine -and $_.CommandLine.Contains($dir) -and $_.CommandLine -like '*uvicorn*'
      }}
    $result = foreach ($p in $procs) {{
      $ports = @(Get-NetTCPConnection -OwningProcess $p.ProcessId -State Listen -ErrorAction SilentlyContinue |
        Select-Object -ExpandProperty LocalPort -Unique)
      [PSCustomObject]@{{ Pid = $p.ProcessId; Ports = $ports }}
    }}
    $result | ConvertTo-Json -Compress
    """
    raw = subprocess.run(
        ["powershell.exe", "-NoProfile", "-NonInteractive", "-Command", ps_script],
        capture_output=True,
        text=True,
        timeout=10,
        check=True,
    ).stdout.strip()

    if not raw:
        return []
    import json

    parsed = json.loads(raw)
    return parsed if isinstance(parsed, list) else [parsed]


def main() -> int:
    try:
        instances = find_running_instances()
    except Exception as err:  # noqa: BLE001 - fail open on any detection error
        print(f"[predev_check] Çalışan instance kontrolü yapılamadı, devam ediliyor: {err}")
        return 0

    if not instances:
        return 0

    print("")
    print("========================================================")
    print(" PostPiloter backend (uvicorn) ZATEN ÇALIŞIYOR.")
    print("========================================================")
    for inst in instances:
        ports = inst.get("Ports") or []
        ports_str = (
            ", ".join(str(p) for p in ports) if ports else "bilinmiyor (henüz dinlemiyor olabilir)"
        )
        print(f"  PID: {inst.get('Pid')}  Port(lar): {ports_str}")
    print("")
    print(" İkinci bir uvicorn süreci başlatmak, aynı porta iki bağımsız")
    print(" sunucunun bağlanmaya çalışmasına ve tutarsız/çakışan davranışa")
    print(" yol açar (daha önce tam olarak bu yaşandı).")
    print("")
    print(" Ne yapmalısınız:")
    print("   - Zaten açık olan terminal/sekmeyi kullanın, YA DA")
    print("   - Eski süreci siz kontrollü şekilde kapatın, örneğin:")
    for inst in instances:
        print(f"       Stop-Process -Id {inst.get('Pid')}")
    print("     (Bu script hiçbir süreci otomatik olarak kapatmaz.)")
    print("========================================================")
    print("")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: apps/frontend/scripts/predev-check.js
#!/usr/bin/env node
/**
 * predev guard: refuses to start a second `next dev` for THIS PostPiloter
 * frontend if one is already running (any port), instead of only checking
 * whether the default port is busy — the real incident that motivated this
 * was a second instance sitting on a different port, invisible to a
 * port-3000-only check.
 *
 * Never kills anything. If detection itself fails, it fails open (allows
 * `next dev` to proceed) rather than blocking a normal start.
 */
const { execFileSync } = require("child_process");
const path = require("path");

const FRONTEND_DIR = path.resolve(__dirname, "..");

function findRunningInstances() {
  // Escape single quotes for the PowerShell string literal.
  const escapedDir = FRONTEND_DIR.replace(/'/g, "''");
  const psScript = `
    $ErrorActionPreference = 'Stop'
    $dir = '${escapedDir}'
    $procs = Get-CimInstance Win32_Process -Filter "Name='node.exe'" |
      Where-Object {
        $_.CommandLine -and $_.CommandLine.Contains($dir) -and (
          $_.CommandLine -like '*next*dev*' -or
          $_.CommandLine -like '*start-server.js*'
        )
      }
    $result = foreach ($p in $procs) {
      $ports = @(Get-NetTCPConnection -OwningProcess $p.ProcessId -State Listen -ErrorAction SilentlyContinue |
        Select-Object -ExpandProperty LocalPort -Unique)
      [PSCustomObject]@{ Pid = $p.ProcessId; Ports = $ports; CommandLine = $p.CommandLine }
    }
    $result | ConvertTo-Json -Compress
  `;

  const raw = execFileSync(
    "powershell.exe",
    ["-NoProfile", "-NonInteractive", "-Command", psScript],
    { encoding: "utf8", timeout: 10000 }
  ).trim();

  if (!raw) return [];
  const parsed = JSON.parse(raw);
  return Array.isArray(parsed) ? parsed : [parsed];
}

function main() {
  let instances;
  try {
    instances = findRunningInstances();
  } catch (err) {
    console.warn(
      "[predev-check] Çalışan instance kontrolü yapılamadı, devam ediliyor:",
      err.message
    );
    return;
  }

  if (instances.length === 0) return;

  console.error("");
  console.error("========================================================");
  console.error(" PostPiloter frontend dev sunucusu ZATEN ÇALIŞIYOR.");
  console.error("========================================================");
  for (const inst of instances) {
    const ports = inst.Ports && inst.Ports.length ? inst.Ports.join(", ") : "bilinmiyor (henüz dinlemiyor olabilir)";
    console.error(`  PID: ${inst.Pid}  Port(lar): ${ports}`);
  }
  console.error("");
  console.error(" İkinci bir `next dev` süreci başlatmak, aynı .next klasörüne");
  console.error(" yazan iki bağımsız derleyici oluşturur ve bozuk chunk/404");
  console.error(" hatalarına yol açar (daha önce tam olarak bu yaşandı).");
  console.error("");
  console.error(" Ne yapmalısınız:");
  console.error("   - Zaten açık olan terminal/sekmeyi kullanın, YA DA");
  console.error("   - Eski süreci siz kontrollü şekilde kapatın, örneğin:");
  for (const inst of instances) {
    console.error(`       Stop-Process -Id ${inst.Pid}`);
  }
  console.error("     (Bu script hiçbir süreci otomatik olarak kapatmaz.)");
  console.error("========================================================");
  console.error("");
  process.exit(1);
}

main();
